group/decaf448/decaf.go

```diff
@@ -0,0 +1,201 @@
+//
+// Decaf Group
+//
+// Decaf (3) is a prime-order group constructed as a quotient of groups. A Decaf
+// element can be represented by any point in the coset P+J[2], where J is a
+// Jacobi quartic curve and J[2] are its 2-torsion points.
+// Since P+J[2] has four points, Decaf specifies rules to choose one canonical
+// representative, which has a unique encoding. Two representations are
+// equivalent if they belong to the same coset.
+//
+// The types Elt and Scalar provide methods to perform arithmetic operations on
+// the Decaf group.
+//
+// Version
+//
+// This implementation uses Decaf v1.0 of the encoding (see (4,5) for a complete
+// specification).
+//
+// References
+//
+// (1) https://www.shiftleft.org/papers/goldilocks
+//
+// (2) https://tools.ietf.org/html/rfc7748
+//
+// (3) https://doi.org/10.1007/978-3-662-47989-6_34 and https://www.shiftleft.org/papers/decaf
+//
+// (4) https://sourceforge.net/p/ed448goldilocks/code/ci/v1.0/tree/
+//
+// (5) https://mailarchive.ietf.org/arch/msg/cfrg/S4YUTt_5eD4kwYbDuhEK0tXT1aM/
+package decaf448
+
+import (
+	"crypto/subtle"
+	"math/bits"
+
+	"github.com/cloudflare/circl/internal/ted448"
+	fp "github.com/cloudflare/circl/math/fp448"
+)
+
+//  Decaf v1.0 of the encoding.
+const Version = "v1.0"
+
+// Elt is an element of the Decaf group. It must be always initialized using
+// one of the Decaf functions.
+type Elt struct{ p ted448.Point }
+
+// Scalar represents a positive integer stored in little-endian order.
+type Scalar = ted448.Scalar
+
+func (e Elt) String() string { return e.p.String() }
+
+// IsValid returns True if a is a valid element of the group.
+func IsValid(a *Elt) bool { return ted448.IsOnCurve(&a.p) }
+
+// Identity returns the identity element of the group.
+func Identity() *Elt { return &Elt{ted448.Identity()} }
+
+// Generator returns the generator element of the group.
+func Generator() *Elt { return &Elt{ted448.Generator()} }
+
+// Order returns a scalar with the order of the group.
+func Order() Scalar { return ted448.Order() }
+
+// Neg calculates c=-a, where - is the inverse of the group operation.
+func Neg(c, a *Elt) { c.p = a.p; c.p.Neg() }
+
+// Add calculates c=a+b, where + is the group operation.
+func Add(c, a, b *Elt) { q := a.p; q.Add(&b.p); c.p = q }
+
+// Double calculates c=a+a, where + is the group operation.
+func Double(c, a *Elt) { c.p = a.p; c.p.Double() }
+
+// Mul calculates c=n*a, where * is scalar multiplication on the group.
+func Mul(c *Elt, n *Scalar, a *Elt) { ted448.ScalarMult(&c.p, n, &a.p) }
+
+// MulGen calculates c=n*g, where * is scalar multiplication on the group,
+// and g is the generator of the group.
+func MulGen(c *Elt, n *Scalar) { ted448.ScalarBaseMult(&c.p, n) }
+
+// IsIdentity returns True if e is the identity of the group.
+func (e *Elt) IsIdentity() bool {
+	b0 := fp.IsZero(&e.p.X)
+	b1 := 1 - fp.IsZero(&e.p.Y)
+	b2 := 1 - fp.IsZero(&e.p.Z)
+	return subtle.ConstantTimeEq(int32(4*b2+2*b1+b0), 0x7) == 1
```

Comment:
Why not simply return `(b0 & b1 & b2)==1`?

Reply:
suggestion accepted